common/src/com/thoughtworks/go/config/materials/PluggableSCMMaterial.java

```diff
@@ -0,0 +1,315 @@
+    }
+
+    @Override
+    public Boolean isUsedInFetchArtifact(PipelineConfig pipelineConfig) {
+        return Boolean.FALSE;
+    }
+
+    @Override
+    public Revision oldestRevision(Modifications modifications) {
+        if (modifications.isEmpty()) {
+            return new NullRevision();
+        }
+        Modification modification = modifications.get(modifications.size() - 1);
+        return new PluggableSCMMaterialRevision(modification.getRevision(), modification.getModifiedTime());
+    }
+
+    @Override
+    public String getLongDescription() {
+        return getUriForDisplay();
+    }
+
+    @Override
+    public void populateEnvironmentContext(EnvironmentVariableContext context, MaterialRevision materialRevision, File workingDir) {
+        context.setProperty(getEnvironmentVariableKey("GO_SCM_%s_%s", "LABEL"), materialRevision.getRevision().getRevision(), false);
+        for (ConfigurationProperty configurationProperty : scmConfig.getConfiguration()) {
+            context.setProperty(getEnvironmentVariableKey("GO_SCM_%s_%s", configurationProperty.getConfigurationKey().getName()),
+                    configurationProperty.getValue(), configurationProperty.isSecure());
+        }
+        HashMap<String, String> additionalData = materialRevision.getLatestModification().getAdditionalDataMap();
+        if (additionalData != null) {
+            for (Map.Entry<String, String> entry : additionalData.entrySet()) {
+                boolean isSecure = false;
+                for (EnvironmentVariableContext.EnvironmentVariable secureEnvironmentVariable : context.getSecureEnvironmentVariables()) {
+                    String urlEncodedValue = null;
+                    try {
+                        urlEncodedValue = URLEncoder.encode(secureEnvironmentVariable.value(), "UTF-8");
+                    } catch (UnsupportedEncodingException e) {
+                    }
+                    boolean isSecureEnvironmentVariableEncoded = !StringUtil.isBlank(urlEncodedValue) && !secureEnvironmentVariable.value().equals(urlEncodedValue);
+                    if (isSecureEnvironmentVariableEncoded && entry.getValue().contains(urlEncodedValue)) {
+                        isSecure = true;
+                        break;
+                    }
+                }
+
+                String key = entry.getKey();
+                String value = entry.getValue();
+                context.setProperty(getEnvironmentVariableKey("GO_SCM_%s_%s", key), value, isSecure);
+            }
+        }
+    }
+
+    private String getEnvironmentVariableKey(String keyPattern, String givenKey) {
+        return escapeEnvironmentVariable(upperCase(format(keyPattern, getName().toString(), givenKey)));
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        // skipping super.equals()
```

Comment:
because super.equals does
```java
        if (name != null ? !name.equals(that.name) : that.name != null) {
            return false;
        }
        if (type != null ? !type.equals(that.type) : that.type != null) {
            return false;
        }
```
type is checked by `getClass() != o.getClass()` while name will be same since its referring to same copy.